config/backend-config/backend-config.go

```diff
@@ -23,52 +25,56 @@ import (
 )
 
 var (
-	backendConfig                         BackendConfig
+	// environment variables
 	configBackendURL                      string
 	pollInterval, regulationsPollInterval time.Duration
-	configFromFile                        bool
 	configJSONPath                        string
-	curSourceJSON                         ConfigT
-	curSourceJSONLock                     sync.RWMutex
-	LastSync                              string
-	LastRegulationSync                    string
+	configFromFile                        bool
 	maxRegulationsPerRequest              int
 	configEnvReplacementEnabled           bool
 
+	LastSync           string
+	LastRegulationSync string
+
 	// DefaultBackendConfig will be initialized be Setup to either a WorkspaceConfig or MultiWorkspaceConfig.
 	DefaultBackendConfig BackendConfig
-	Http                 sysUtils.HttpI   = sysUtils.NewHttp()
 	pkgLogger            logger.LoggerI   = logger.NewLogger().Child("backend-config")
 	IoUtil               sysUtils.IoUtilI = sysUtils.NewIoUtil()
 	Diagnostics          diagnostics.DiagnosticsI
 )
 
-type BackendConfig interface {
+type workspaceConfig interface {
 	SetUp() error
 	AccessToken() string
 	Get(context.Context, string) (ConfigT, error)
 	GetWorkspaceIDForWriteKey(string) string
 	GetWorkspaceIDForSourceID(string) string
 	GetWorkspaceLibrariesForWorkspaceID(string) LibrariesT
+}
+
+type BackendConfig interface {
+	workspaceConfig
 	WaitForConfig(ctx context.Context)
 	Subscribe(ctx context.Context, topic Topic) pubsub.DataChannel
 	Stop()
 	StartWithIDs(ctx context.Context, workspaces string)
+	GetConfig() ConfigT
 }
 
-type CommonBackendConfig struct {
-	eb               *pubsub.PublishSubscriber
-	configEnvHandler types.ConfigEnvI
-	ctx              context.Context
-	cancel           context.CancelFunc
-	blockChan        chan struct{}
-	initializedLock  sync.RWMutex
-	initialized      bool
+type commonBackendConfig struct {
+	workspaceConfig   workspaceConfig
+	eb                *pubsub.PublishSubscriber
+	ctx               context.Context
+	cancel            context.CancelFunc
+	blockChan         chan struct{}
+	initializedLock   sync.RWMutex
+	initialized       bool
+	curSourceJSON     ConfigT
+	curSourceJSONLock sync.RWMutex
```

Comment:
moving the lock above the resource it protects is preferred

Reply:
where did you get this? as long as they are 1 line closed to each other it's cool imho :thinking:

Reply:
It seems to be a typical pattern in Go's own repository, e.g [here](https://github.com/golang/go/blob/2580d0e08d5e9f979b943758d3c49877fb2324cb/src/net/rpc/jsonrpc/server.go#L31), [here](https://github.com/golang/go/blob/2580d0e08d5e9f979b943758d3c49877fb2324cb/src/syscall/js/func.go#L12) & [here](https://github.com/golang/go/blob/2ebe77a2fda1ee9ff6fd9a3e08933ad1ebaea039/src/net/textproto/pipeline.go#L29), unless it is a test where they seem to be more relaxed, e.g. [here](https://github.com/golang/go/blob/d6d2ebb7b8194eb20d95b84bae39f7a5837e9f72/test/typeparam/genembed2.go#L22).
I remember reading it initially in some article and it made sense. I don't remember the exact article, but I could find a similar one [here](https://medium.com/hackernoon/dancing-with-go-s-mutexes-92407ae927bf#3426) as well.

Reply:
Interesting stuff, thanks. Not sure I feel the need though :man_shrugging: